<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <link href="//fonts.googleapis.com/css?family=Raleway:400,300,600" rel="stylesheet" type="text/css">
  <link rel="stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/skeleton/2.0.4/skeleton.css"/>

  <!-- Necessary includes for LocusZoom.js -->
 <script src="https://cdn.jsdelivr.net/npm/d3@^5.16.0" type="text/javascript"></script>
  <script src="../dist/locuszoom.app.min.js" type="text/javascript"></script>
  <link rel="stylesheet" href="../dist/locuszoom.css" type="text/css"/>

  <script type="application/javascript" src="../dist/ext/lz-dynamic-urls.min.js"></script>

  <title>LocusZoom.js ~ GWAS Catalog</title>

  <style>
    body {
      background-color: #FAFAFA;
      margin: 0 20px;
    }

    img {
      max-width: 100%;
      box-sizing: border-box;
    }

    ul.top_hits li {
      margin-bottom: 0;
    }
  </style>

</head>

<body>
<div class="container">

  <h1 style="margin-top: 1em;"><strong>LocusZoom.js</strong></h1>

  <h3 style="float: left; color: #777">GWAS Catalog Demonstration</h3>
  <h6 style="float: right;"><a href="../index.html">&lt; return home</a></h6>

  <hr style="clear: both;">

  <p>
    It can be helpful to view an association plot in context of all previously reported claims about an association,
    across all traits and studies. LocusZoom is able to mark SNPs that have been previously reported in any of several
    GWAS catalogs, including the NHGRI-EBI GWAS catalog (build 37) used in this demonstration.
  </p>

  <p>
    Significant association SNPs with entries in the catalog can be marked with labels on the association plot, a separate
    annotation "rug" track, or both. In the case where the GWAS Catalog contains multiple reports about a single SNP,
    all entries are matched by chromosome and position, and the trait with the most significant pvalue is reported.
    The rug track provides more detailed contextual information.
  </p>
  <p>
    In this demonstration, trait labels can be toggled for GWAS Catalog SNPs within LD > 0.4 of the reference variant.
    To begin, move your mouse over the association results plot below and select "Display Options".
  </p>

  <h3>Interactive Plot</h3>
  <div class="row">
    <div class="two columns">
      <h4>Top Hits</h4>
      <ul class="top_hits" style="padding-left: 0.2rem; min-width: 110px;"></ul>
    </div>
    <div class="ten columns">
      <div id="lz-plot" class="lz-container-responsive"></div>
    </div>
  </div>
  <hr>

  <div class="row">
    <footer style="text-align: center;">
      &copy; Copyright <script>document.write(new Date().getFullYear())</script> <a href="https://github.com/statgen">The University of Michigan Center for Statistical
      Genetics</a><br>
    </footer>
  </div>

  <script type="application/javascript">
      /*
        Specify the data sources to use
       */
      var apiBase = "//portaldev.sph.umich.edu/api/v1/";
      var data_sources = new LocusZoom.DataSources()
          .add("assoc", ["AssociationLZ", { url: apiBase + "statistic/single/", params: { source: 45, id_field: "variant" } }])
          .add("ld", ["LDLZ2", { url: "https://portaldev.sph.umich.edu/ld/", params: { source: '1000G', population: 'ALL' } }])
          .add("gene", ["GeneLZ", { url: apiBase + "annotation/genes/" }])
          .add("catalog", ["GwasCatalogLZ", { url: apiBase + 'annotation/gwascatalog/results/' }])
          .add("recomb", ["RecombLZ", { url: apiBase + "annotation/recomb/results/" }])
          .add("constraint", ["GeneConstraintLZ", { url: "https://gnomad.broadinstitute.org/api", params: { build: 'GRCh37' } }]);

      /*
        Define and render the plot
       */
      // Fetch custom layout defined for usage with credible sets
      var stateUrlMapping = {chr: "chrom", start: "start", end: "end"};
      // Fetch initial position from the URL, or use some defaults
      var initialState = LzDynamicUrls.paramsFromUrl(stateUrlMapping);
      if (!Object.keys(initialState).length) {
        initialState = { chr: 9, start: 21751670, end: 22351670 };
      }
      // This example demonstrates using the `state.genome_build` param, which tells LZ annotation sources to
      //  automatically select an appropriate dataset based on a globally specified genome build.
      initialState.genome_build = 'GRCh37';

      layout = LocusZoom.Layouts.get("plot", "association_catalog", {
          state: initialState
      });
      // Although the demo will work very well with the base layout, this section adds some additional custom features
      var layout_assoc_layer = layout.panels[1].data_layers[2];
      layout_assoc_layer.match = { send: 'assoc:position', receive: 'assoc:position' };
      layout_assoc_layer.color.unshift({
          field: 'lz_highlight_match',  // Special field name whose presence triggers custom rendering
          scale_function: 'if',
          parameters: {
              field_value: true,
              then: '#FFF000'
          }
      });

      var layout_annotation_track = layout.panels[0].data_layers[0];
      layout_annotation_track.match = { send: 'catalog:pos', receive: 'catalog:pos' };
      layout_annotation_track.color = [
          {
              field: 'lz_highlight_match',  // Special field name whose presence triggers custom rendering
              scale_function: 'if',
              parameters: {
                  field_value: true,
                  then: '#FF0000'
              }
          },
          '#0000CC'
      ];

      // Allow users to select custom LD population. This button isn't part of the builtin LocusZoom layouts because not
      //  everyone uses the UM 1000G LDServer (LDLZ2 datasource)
      layout.toolbar = LocusZoom.Layouts.get("toolbar", "region_nav_plot");
      layout.toolbar.widgets.push(LocusZoom.Layouts.get('toolbar_widgets', 'ldlz2_pop_selector'));

      // Generate the LocusZoom plot
      window.plot = LocusZoom.populate("#lz-plot", data_sources, layout);

      // Changes in the plot can be reflected in the URL, and vice versa (eg browser back button can go back to
      //   a previously viewed region)
      LzDynamicUrls.plotUpdatesUrl(plot, stateUrlMapping);
      LzDynamicUrls.plotWatchesUrl(plot, stateUrlMapping);


      /*
        Add a sidebar to the page with "top hits" items

        RBMS1, for example, has a region of high LD and therefore includes many points in the credible set
       */
      // Load a region of the plot based on a specified region string
      function jumpTo(region) {
          var target = region.split(":");
          var chr = target[0];
          var pos = target[1];
          var start = 0;
          var end = 0;
          if (!pos.match(/[-+]/)) {
              start = +pos - 300000;
              end = +pos + 300000;
          }
          plot.applyState({chr: chr, start: start, end: end, ldrefvar: ""});
          return false;
      }

      // Populate a list of top hits links for the plot
      var top_hits = [
          ["9:22051670", "CDKN2A/B (default)"],
          ["7:15052860", "DGKB"],
          ["16:75247245", "BCAR1"],
          ["15:77832762", "HMG20A"],
          ["8:95937502", "TP53INP1"],
          ["10:114758349", "TCF7L2"],
          ["16:53819169", "FTO"],
          ["7:28196413", "JAZF1"],
          ["12:71433293", "TSPAN8"],
          ["6:20679709", "CDKAL1"],
          ["2:161346447", "RBMS1"]
      ];
      top_hits.forEach(function (hit) {
          d3.select("ul.top_hits").append("li")
              .html("<a href=\"javascript:void(0);\" onclick=\"javascript:jumpTo('" + hit[0] + "');\">" + hit[1] + "</a>");
      });
  </script>

</div>

</body>
</html>
